utils/data-ingestor.js

```diff
@@ -0,0 +1,229 @@
+   * based on the type of 'special' row. Additionally, scales the recomputed sum value
+   * if configured in 'options'
+   * @param{Row} row - The row to operate on
+   * @param{string} specialType - The type of this special variable
+   * @param{string} variable - The variable for the given row
+   * @param{string} year - The year for the given dataset
+   * @param{Object} options - Additional configuration for special calculations
+   * @returns{Row}
+   */
+  recomputeSum(row, specialType, variable, year, options) {
+    let updatedRow = row;
+    let sum;
+
+    if (specialType === 'median') {
+      const { trimmedEstimate, codingThreshold } = interpolate(this.data, variable, year);
+      sum = trimmedEstimate;
+      if (codingThreshold) {
+        updatedRow = updatedRow.set('codingThreshold', codingThreshold);
+        // if codingThreshold is set, indicates value was top- or bottom-coded,
+        // meaning the value is not reliable
+        updatedRow = updatedRow.set('is_reliable', false);
+      }
+    } else {
+      const formulaName = getFormulaName(options, 'sum');
+      sum = executeFormula(this.data, variable, formulaName, options.args);
+    }
+    sum = this.applyTransform(sum, options.transform);
+
+    return updatedRow.set('sum', sum);
+  }
+
+  /*
+   * Recomputes the 'm'(margin of error) value for a given row, applying the appropriate calculation
+   * based on the type of 'special' row. Additionally, scales the recomputed m value
+   * if configured in 'options'
+   * @param{Row} row - The row to operate on
+   * @param{string} specialType - The type of this special variable
+   * @param{string} variable - The variable for the given row
+   * @param{string} year - The year for the given dataset
+   * @param{Object} options - Additional configuration for special calculations
+   * @returns{Row}
+   */
+  recomputeM(row, specialType, variable, year, options) {
+    const updatedRow = row;
+    let m;
+    if (specialType === 'median') {
+      m = calculateMedianError(this.data, variable, year, options);
+    } else {
+      const formulaName = getFormulaName(options, 'm');
+      m = executeFormula(formulaName, options.args, this.data, variable);
+    }
+
+    m = this.applyTransform(m, options.transform);
+
+    return updatedRow.set('m', m);
+  }
+
+  /*
+   * If transformOptions exists, then applies the appropriate transformation.
+   * 'inflate' is a special transform type, which only applies to previous year data points
+   * (designated by this.isPrevious), and scales by INFLATION_FACTOR. Other allowed transformation
+   * is 'scale' type, and require transformOptions.factor; factor is applied as scalar transformation
+   * @param{Number} val - The value to transform
+   * @param{Object} [transformOptions] - Object containing 'type', and optionally 'factor'; defines the transformation
+   * @returns{Number}
+   */
+  applyTransform(val, transformOptions) {
+    if (!transformOptions) return val;
```

Comment:
I really like this - the methods themselves contain their own validation logic, as they should. I sometimes validate on the calling side, which mixes up concerns across methods!!!